File: app/src/main/java/mobrsg/android/redsinergia/gallerydemo/api/RetrofitAdapter.java
package mobrsg.android.redsinergia.gallerydemo.api;

import java.security.Key;
import java.util.Arrays;
import java.util.Collections;

import okhttp3.CertificatePinner;
import okhttp3.CipherSuite;
import okhttp3.ConnectionSpec;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.TlsVersion;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static mobrsg.android.redsinergia.gallerydemo.Constants.BASE_URL;
import static mobrsg.android.redsinergia.gallerydemo.Constants.KEY;

/*
 * Crea la instancia de retrofit para despues crear el request a la API
 * el addInterceptor sirve para agregar header a tu peticionen caso de que lo requiera como
 * en este caso se requiere un Authorization y loa asignamos aqui
 * tambien puedes agregar un timeout en esta parte
 * */

public class RetrofitAdapter {

    public Retrofit getInstance() {

        ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS)
                .tlsVersions(TlsVersion.TLS_1_2)
                .cipherSuites(
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_DSS_WITH_AES_128_CBC_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA)
                .build();

        OkHttpClient client = new OkHttpClient.Builder()
                .connectionSpecs(Collections.singletonList(spec))
                .addInterceptor(chain -> {
                    Request request = chain.request();
                    request = request.newBuilder()
                            .addHeader("Authorization", "Client-ID " + KEY)
                            .build();
                    return chain.proceed(request);
                })
                .build();

        return new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

}

File: app/src/main/java/mobrsg/android/redsinergia/gallerydemo/api/RetrofitAdapter.kt
package mobrsg.android.redsinergia.gallerydemo.api

import mobrsg.android.redsinergia.gallerydemo.Constants.BASE_URL
import mobrsg.android.redsinergia.gallerydemo.Constants.KEY
import okhttp3.Interceptor.Companion.invoke
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import okhttp3.OkHttpClient

class RetrofitAdapter {

    fun getInstance(): Retrofit {

        val logging = HttpLoggingInterceptor()
        logging.setLevel(HttpLoggingInterceptor.Level.BODY)

        val client = OkHttpClient.Builder()
                .addInterceptor(logging)
                .addInterceptor(invoke {
                    var request = it.request()
                    request = request.newBuilder()
                            .addHeader("Authorization", "Client-ID $KEY")
                            .build()
                    it.proceed(request)
                })
                .build()

        return Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

}